samples/tutorial/Tutorial3.md

```diff
@@ -0,0 +1,415 @@
+  )
+}
+```
+
+This view isn't particularly useful without the data to present it, so update the `TodoEditScreen` to add the properties we need and the callbacks:
+
+```kotlin
+data class TodoEditScreen(
+  /** The title of this todo item. */
+  val title: String,
+  /** The contents, or "note" of the todo. */
+  val note: String,
+
+  /** Callbacks for when the title or note changes. */
+  val onTitleChanged: (String) -> Unit,
+  val onNoteChanged: (String) -> Unit,
+
+  val discardChanges: () -> Unit,
+  val saveChanges: () -> Unit
+)
+```
+
+Then update the view with the data from the screen:
+
+```kotlin
+class TodoEditLayoutRunner(
+  private val binding: TodoEditViewBinding
+) : LayoutRunner<TodoEditScreen> {
+
+  override fun showRendering(
+    rendering: TodoEditScreen,
+    viewEnvironment: ViewEnvironment
+  ) {
+    binding.root.backPressedHandler = rendering.discardChanges
+    binding.save.setOnClickListener { rendering.saveChanges() }
+    binding.todoTitle.updateText(rendering.title)
+    binding.todoTitle.setTextChangedListener { rendering.onTitleChanged(it.toString()) }
+    binding.todoNote.updateText(rendering.note)
+    binding.todoNote.setTextChangedListener { rendering.onNoteChanged(it.toString()) }
+  }
+
+  // …
+}
+```
+
+#### TodoEditWorkflow
+
+Now that we have our screen and view controller, let's update the `TodoEditWorkflow` to emit this screen as the rendering.
+
+The `TodoEditWorkflow` needs an initial Todo item passed into it from its parent. It will make a copy of it in its internal state (because `TodoModel` is a value type) — this can be the "scratch pad" for edits. This allows changes to be made and still be able to discard the changes if the user does not want to save them.
+
+Additionally, we will (finally) use the `onPropsChanged` method. If the edit workflow's parent provides an updated `todo`, it will invalidate the `todo` in `State`, and replace it with the one provided from the props.
+
+```kotlin
+object TodoEditWorkflow : StatefulWorkflow<EditProps, State, Output, TodoEditScreen>() {
+
+  data class EditProps(
+    /** The "Todo" passed from our parent. */
+    val initialTodo: TodoModel
+  )
+
+  data class State(
+    /** The workflow's copy of the Todo item. Changes are local to this workflow. */
+    val todo: TodoModel
+  )
+
+  sealed class Output {
+  }
+
+  override fun initialState(
+    props: EditProps,
+    snapshot: Snapshot?
+  ): State = State(props.initialTodo)
+
+  override fun onPropsChanged(
+    old: EditProps,
+    new: EditProps,
+    state: State
+  ): State {
+    // The `Todo` from our parent changed. Update our internal copy so we are starting from the same
+    // item. The "correct" behavior depends on the business logic - would we only want to update if
+    // the users hasn't changed the todo from the initial one? Or is it ok to delete whatever edits
+    // were in progress if the state from the parent changes?
+    if (old.initialTodo != new.initialTodo) {
```

Comment:
Isn't this check redundant these days?

Reply:
Yes, although this doesn't hurt since if other properties were added to the props, we'd probably have to add this back in.